### PR TITLE
[atspi-common] Opaque macro split-up for clearer macros

### PR DESCRIPTION
**Pull request by luukvanderduim**
`impl_event_conversions` expands to five conversions, depending on its argument count. 
It is not clear at the call site to which conversions the macro will expand and it is unknown what arguments are expected at that point.
    
This PR introduces five separate macros and documents each,
such that, when called, it will be known which arguments are required and
what it will expand to.
Also the commits aim to make clear, by name, what each of the macros is shorthand for.
    
Upside:
- improve readability by knowing what the macro is short-hand for,
- improve debug-ability as each macro now expands to one conversion only.

Downside:    
This comes at the cost of being a bit more verbose.

Introduces the following macros:
    
    1. `impl_from_interface_event_enum_for_event`
    2. `impl_try_from_event_for_user_facing_event_type`
    3. `impl_from_user_facing_event_for_interface_event_enum`
    4. `impl_from_user_facing_type_for_event_enum`
    5. `impl_try_from_event_for_user_facing_type`
    
First commit introduces the macros,
Second commit replaces instances with two arguments
Third commit replaces instances with four arguments
Last commit removes `impl_event_conversions`

### CODE REVIEW

**Comment by TTWNO** [comment]
This change is fine. I get a little bit excited about macros so I guess I went a bit overboard.

I think you're correct about readability. To me, this is similar to moving away from auto-generated types and using defined, concrete types, then testing against the XML as part of CI. This seems much more reasonable.

**Comment by TTWNO** [comment]
Would you also like to do the same with the `event_test_cases!`? It expands to three macros, with some additional bounds checking.

Personally, I'd like to see that macro stay nested, since that would make adding new types of tests extremely tedious.

File: atspi-common/src/macros.rs [hunk 1]
```diff
@@ -22,17 +58,86 @@ macro_rules! impl_event_conversions {
...
+/// ```ignore
+/// impl TryFrom<Event> for StateChangedEvent {
+///    type Error = AtspiError;
+///   fn try_from(generic_event: Event) -> Result<StateChangedEvent, Self::Error> {
+///      if let Event::Object(ObjectEvents::StateChanged(specific_event)) = generic_event {
+///          Ok(specific_event)
+///         } else {
+///          Err(AtspiError::Conversion("Invalid type"))
+///         }
+/// }
```
[the remark this answers is not in the document]
Comment (luukvanderduim) [reply atspi-common/src/macros.rs:137]: Good catch. thanks. 

Fixed.

**Comment by TTWNO** [comment]
What is the reasoning behind the `ignore` tags on the documentation examples? Leveraging correctness of the examples would make the documentation a higher quality, would it not?

I'm just checking if there's a reason behind that decision.

**Comment by TTWNO** [comment]
Oh, I'm silly. You can't implement traits twice for a single struct. Got it.

**Comment by TTWNO** [comment]
Great work, Luuk! The docs are much, much improved thanks to you!

**Review by TTWNO — APPROVED** [review]

**Pull request merged**